### ISSUE TITLE
SDL_CoreAudio: Use correct 7.1 audio configuration

### ISSUE THREAD

**Issue by Danos0100** [libsdl-org/SDL#11709]
At the moment, when receiving an 8 channel stream, sdl is harcoded to use kAudioChannelLayoutTag_MPEG_7_1_A as channel layout. This layout however is not commonly used, as it uses a 'left of center' and a 'right of center' speaker between Front Left and Center/between Front Right and Center.

https://github.com/libsdl-org/SDL/blob/e79b0ce2e4660e7621719ca3fce362545c90cbf7/src/audio/coreaudio/SDL_coreaudio.m#L882
(L R C LFE Ls Rs Lc Rc)
This leads to SDL-apps on iOS/tvOS trying to output 7.1 audio (like steam link for example) to actually play back surround back audio stream on the surround side speakers and the surround side stream on the front left + center and front right + center speakers, while the surround back speakers stay silent. This also applies to virtualized spatial sound when using AirPods and letting apple do the spatialization.

This needs to be changed to
kAudioChannelLayoutTag_WAVE_7_1 (8 channels, L R C LFE Rls Rrs Ls Rs) (this is the same channel order that is used in windows).

SDL3 has the same problem:
https://github.com/libsdl-org/SDL/blob/f40a530fd9fcdea79986de12d305426c0304d60e/src/audio/coreaudio/SDL_coreaudio.m#L781

(credit goes to @andygrundman for helping me to find out, why moonlight-ios and steam link use the 7.1 front layout)

(also tagging @slouken as this breaks steam link's 7.1 sound on iOS and tvOS currently)

**Comment by slouken** [comment]
Fixed, thanks!

**Comment by barracuda156** [comment]
@Danos0100 @slouken @@andygrundman These may not exist, which leads to a broken build:
```
/opt/local/var/macports/build/_opt_PPCSnowLeopardPorts_devel_libsdl2-powerpc/libsdl2-powerpc/work/SDL2-2.30.11/src/audio/coreaudio/SDL_coreaudio.m:894:36: error: 'kAudioChannelLayoutTag_WAVE_7_1' undeclared (first use in this function); did you mean 'kAudioChannelLayoutTag_AAC_7_1'?
  894 |         layout.mChannelLayoutTag = kAudioChannelLayoutTag_WAVE_7_1;
      |                                    ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
      |                                    kAudioChannelLayoutTag_AAC_7_1
```

Perhaps worth checking what is available instead or making iOS-specific fix specifically for iOS?

**Comment by Danos0100** [comment]
Strange, build worked fine for me. This is already apple specific, as CoreAudio is an Apple Framework that is only used on iOS/iPadOS/VisionOS/tvOS/macOS.

`kAudioChannelLayoutTag_AAC_7_1` is identical to `kAudioChannelLayoutTag_MPEG_7_1_B` and thus a  C Lc Rc L R Ls Rs Lfe configuration (with Left Center and Right Center – nothing we want!)

**Comment by barracuda156** [comment]
@Danos0100 Is there a fallback for WAVE ones? Those are breaking, the rest work.

**Comment by andygrundman** [comment]
According to Apple [this has been around](https://developer.apple.com/documentation/coreaudiotypes/kaudiochannellayouttag_wave_7_1?changes=_1) since the iOS 13 SDK, which is actually not as long as I would have thought.  I think the WAVE layouts are named that way because they come from the [Microsoft standard layouts](https://www.jensign.com/bdp95/7dot1voiced/index.html) used in Windows and WAV files. There are so many unusual layouts that we felt that using the Microsoft layout was a good default. Our primary use case here is the Moonlight client streaming audio from a Windows machine, where this is the only 7.1 layout.

Rls/Rrs = Rear left/right surround aka back surround
```
kAudioChannelLayoutTag_WAVE_7_1     = (189U<<16) | 8,       ///< 8 channels, L R C LFE Rls Rrs Ls Rs
```

I just noticed in SDL_audio.h that this is also the 7.1 layout mentioned as the SDL default. That seems like a solid indication that the MPEG layouts in SDL_coreaudio should be changed to match. In addition to 7.1 these are wrong, but are probably too obscure to worry about.
```
kAudioChannelLayoutTag_MPEG_5_0_A    = (117U<<16) | 5,      ///<  L R C Ls Rs
kAudioChannelLayoutTag_MPEG_6_1_A    = (125U<<16) | 7,      ///<  L R C LFE Ls Rs Cs
```
```
/**
 * For multi-channel audio, the default SDL channel mapping is:
 *
 * 2:  FL  FR                          (stereo)
 * 3:  FL  FR LFE                      (2.1 surround)
 * 4:  FL  FR  BL  BR                  (quad)
 * 5:  FL  FR LFE  BL  BR              (4.1 surround)
 * 6:  FL  FR  FC LFE  SL  SR          (5.1 surround - last two can also be BL BR)
 * 7:  FL  FR  FC LFE  BC  SL  SR      (6.1 surround)
 * 8:  FL  FR  FC LFE  BL  BR  SL  SR  (7.1 surround)
 */ 
```

**Comment by barracuda156** [comment]
@andygrundman Thank you for a detailed response. If there are no equivalents of `kAudioChannelLayoutTag_WAVE_*` with a better compatibility to fall back to, could you advise what I rather use in place of those on a version of macOS where they are not defined? I have no concern for supporting iOS (I mean, locally I do not need it). Obviously, I can just replace these two with what was used prior to the breaking change, but maybe there is a better option?

**Comment by slouken** [comment]
@barracuda156, the build error was fixed by @jmroot for the next release in 81ce8d659ede1ffcb8e713020d32e7cfb6244689.

There's an open FIXME in the SDL3 code to swizzle the channels on older OS instead of reporting that 7.1 surround sound is not available. Given that it only affects macOS 10.14 and earlier and iOS 12 and earlier, it's likely to need an external PR to be fixed.

**Comment by barracuda156** [comment]
@slouken Great, and thank you for the reference. I will borrow the fix.

@jmroot Thanks for addressing this!
